### ISSUE TITLE
Can't add a VIM without Mano Provider's Sync on

### ISSUE THREAD

**Issue by DanielGomes14** [openslice/io.openslice.portal.api#18]
I have recently added a ManoProvider without the Sync ability on which obliges me to create a VIM.  However, when adding a VIM I have faced some problems. More precisely,  in the line 3363 of the ArtifactsAPIController:

```java
@PostMapping(value = "/admin/infrastructures", produces = "application/json", consumes = "application/json")
	public ResponseEntity<?> addInfrastructure(@Valid @RequestBody Infrastructure c, HttpServletRequest request)
			throws ForbiddenException {
		
		Object attr = request.getSession().getAttribute("SPRING_SECURITY_CONTEXT");
		SecurityContextHolder.setContext((SecurityContext) attr);

		if (!checkUserIDorIsAdmin(-1)) {
			throw new ForbiddenException("The requested page is forbidden");// return (ResponseEntity<?>)
																			// ResponseEntity.status(HttpStatus.FORBIDDEN)
																			// ;
		}
		Infrastructure u = infrastructureService.addInfrastructure(c);

		if (u != null) {
			return ResponseEntity.ok(u);
		} else {
			return (ResponseEntity<?>) ResponseEntity.badRequest().build();
		}
	}
```
The `SecurityContextHolder` is unable to set the context to value of the `attr`Object  since it is null, throwing an Exception. These conclusions are supported by the following logs:
```
java.lang.IllegalArgumentException: Only non-null SecurityContext instances are permitted
09:27:53.643 [http-nio-13000-exec-8] ERROR o.a.c.c.C.[.[.[.[dispatcherServlet] - Servlet.service() for servlet [dispatcherServlet] in context with path [/osapi] threw exception [Request processing failed: java.lang.IllegalArgumentException: Only non-null SecurityContext instances are permitted] with root cause
	at org.springframework.util.Assert.notNull(Assert.java:204)
	at org.springframework.security.core.context.ThreadLocalSecurityContextHolderStrategy.setContext(ThreadLocalSecurityContextHolderStrategy.java:59)
	at org.springframework.security.core.context.SecurityContextHolder.setContext(SecurityContextHolder.java:155)
	at portal.api.controller.ArtifactsAPIController.addInfrastructure(ArtifactsAPIController.java:3363)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
```

I have tackled these issue by commenting the lines 3362 and 3363 of the ArtifactsAPIController. However, it led me to another issue. For instance, the status of the VIM is Unknown and therefore unable to attach to any ManoProvider available. This issue is depicted in the following figure.
![Vim Error](https://github.com/openslice/io.openslice.portal.api/assets/48451636/42586e4e-dc98-48f8-89ca-a9cabfcb3ea0)

After looking into the code, I have realised that the update of the VIM status from UKNOWN to OSM_PRESENT relies on the [synchronizeVIMs](https://github.com/openslice/io.openslice.mano/blob/3869a6cd5138eedca848edb58895d6dd04aaf4f4/src/main/java/io/openslice/mano/MANOController.java#L944C76-L944C76) method of the `io.openslice.mano`module , which is only triggered when the Mano Provider Sync is on. 

Would it be possible to attach MANO Providers when adding a VIM?

**Comment by ctranoris** [comment]
I cc @Ioannischatzis who might help.
otherwise I will check also later

**Comment by ctranoris** [comment]
I'm afraid these lines affect many places since the new migration
Object attr = request.getSession().getAttribute("SPRING_SECURITY_CONTEXT");
		SecurityContextHolder.setContext((SecurityContext) attr);
for example do we accept new deployments on endpoint /admin/deployment?

**Comment by ctranoris** [comment]
part of the security issue in #19

**Comment by Ioannischatzis** [comment]
I believe the issue is not relevant to the sync feature. By inspecting the normal flow of the code, the checkUserIDorIsAdmin(long userID) checks if the current user has the ADMIN role or if the user logged in is equal to the requested id (which in this case is set to -1 by the code to skip this check in this specific case).
Thus, in order for a user to create a VIM, it is needed to have the admin role. Does the user have the ADMIN role?
In order to do that a change in the keycloak configuration is needed to assign the user to the ADMIN role. This procedure is described in the documentation of the installation. Please check https://openslice.readthedocs.io/en/stable/deployment/, search for "Post installation steps" and do the procedure mentioned to add to the Assigned Roles the ADMIN role for the user.

**Comment by DanielGomes14** [comment]
Yes, the user has the ADMIN role. However the main point of this issue is not over the security aspect. As I mentioned, i tackled this by commenting the lines 3362 and 3363. By doing so, I was able to a create a VIM. However, this VIM is not attached to any MANO Provider, being therefore in the Infrastructure State "UNKNOWN".  From my understanding, the only update from this state to "OSM_PRESENT" is in [synchronizeVIMs](https://github.com/openslice/io.openslice.mano/blob/3869a6cd5138eedca848edb58895d6dd04aaf4f4/src/main/java/io/openslice/mano/MANOController.java#L944C76-L944C76) method, used for Mano Providers Syncing.
I think it should be possible to add a valid VIM without having to have the sync feature enabled.

**Comment by Ioannischatzis** [comment]
As the code implies as the communication with an OSM is not established the VIM is marked as uknown.
In order to understand what this affects in order to create a fix, can I ask what would be the use case affected by this? In which case the VIM should not be required to be present in an OSM in order to instantiate a deployment?

**Comment by DanielGomes14** [comment]
I want to instantiate a deployment. However I do not want to have the sync feature enabled, as the NFV portal would fetch all the artifacts from my OSM. From what I have checked, it is not possible to instantiate a deployment in these conditions, since the VIM added manually has not been recognized/attached by any of my mano providers 🙂 .

**Comment by Ioannischatzis** [comment]
Have you tried to instantiate a deployment without the sync enabled even if the VIM is in UKNOWN status? You should provide the VIM id and associate the VIM with the corresponding OSM manually to the portal.

**Comment by DanielGomes14** [comment]
I provided the correct VIM id, however how can I associate the VIM with the corresponding OSM manually to the portal? I have looked and I haven't able to find where I can i do so.

**Comment by Ioannischatzis** [comment]
You can go from the menu Admin->Manage MANO Providers. There you can add a MANO provider (OSM). Then go to menu Admin->Manage Infrastructures and add a new VIM and in Datacentername use the Name of the MANO provider you created. The Infrastructures Datacentername is the foreign key that maps the OSM with the corresponding VIM.

**Comment by DanielGomes14** [comment]
I have tried to do so and it failed. The Infrastructure name and Organization are matching the same name of the VIM registered on OSM, while the VIM ID is the same as the one in OSM. Once i tried to import a NSD as a Service Specification in the Services Portal ( as part of the process to order a Service Order), it throws an internal Server Error due to this line in the TMF'S API  [specFromNSDID](https://github.com/openslice/io.openslice.tmf.api/blob/65e7fe8d4ab68313d4413ad4d7be1a5cc0408435/src/main/java/io/openslice/tmf/scm633/reposervices/ServiceSpecificationRepoService.java#L1181) method. As one may see, since the manoprovider in cause does not have a VIM associated, it results in an error. This error is shown in the following logs:
```
15:07:42.782 [http-nio-13082-exec-5] ERROR o.a.c.c.C.[.[.[.[dispatcherServlet] - Servlet.service() for servlet [dispatcherServlet] in context with path [/tmf-api] threw exception [Request processing failed: java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0] with root cause
java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
	at io.openslice.tmf.scm633.reposervices.ServiceSpecificationRepoService.specFromNSDID(ServiceSpecificationRepoService.java:1181)
	at java.base/java.util.ArrayList.get(ArrayList.java:427)
	at java.base/java.util.Objects.checkIndex(Objects.java:361)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:266)
```

Thus, I think that the problems relies in the Portal API, as the VIM is not being associated with the corresponding OSM when adding manually to the portal.

**Comment by Ioannischatzis** [comment]
You probably need to onboard the vnfd and nsd through the nfv portal first and then proceed to the service specification as you described.

**Comment by kostistr** [comment]
Hello @DanielGomes14, as it seems, the manual addition of a VIM to a MANO provider does not work as intended. 

Therefore the import of an NSD as a Service Specification will always fail. 

The issue is located entirely at the portal API and portal web project. We will get back as soon as we have fully replicated and examined the issue.

**Comment by DanielGomes14** [comment]
Hi @kostistr, thank you for the feedback. As a suggestion to fix this issue, I think that adding a field in the VIM creation that allows the association of the MANO Providers that the user wants to the VIM being added. If you agree with this suggestion, I can implement this feature 🙂 .

**Comment by kostistr** [comment]
It will need more thorough investigation because the association of MANO Provider with its VIMs is not only referenced in the VIM creation/edit UI, but in many entities also, such as the MANO Provider itself, the onboarding descriptor, even in the NFV artifacts themselves (hence the error you are receiving, when trying to deploy a given NS).

**Comment by Ioannischatzis** [comment]
Now you can add the OSM's name in the filed named "VIM's MANO Provider Name" and the infrastructure will be mapped to the OSM with the same name. Please get also the latest changes of the portal.web project which renames the Datacenter name to "VIM's MANO Provider Name"

**Comment by DanielGomes14** [comment]
Thank you for the fix @Ioannischatzis ! However there is one issue with your approach: as the VIM's MANO Provider Name is not unique, there might occur a mismatch if there are more than one MANO Providers with the same Name. For instance, it would select the first entry which perhaps, might not the one desired by the user. Thus, I updated you fix, through the PR's [#6](https://github.com/openslice/io.openslice.portal.web/pull/6) in portal.web and [#21](https://github.com/openslice/io.openslice.portal.api/pull/21) in portal.api.  Consequently, the user is now able to choose the mano provider to be associated with the infrastructure and, in turn, its' name will be linked to the VIM DataCenter name.  In the endpoint addInfrastructure and editInfrastructure, we only need to verify if the ManoProvider object has been included in the payload of the create/edit Infrastructure.  It should be mentioned that no models have been affected, maintaining the core structure of the relationships in the database.
